### ISSUE TITLE
[java] UnnecessaryBoxing, but explicit conversion is necessary

### ISSUE THREAD

**Issue by andygoossens** [pmd/pmd#4903]
**Affects PMD Version:** 7.0.0

**Rule:** [UnnecessaryBoxing](https://docs.pmd-code.org/latest/pmd_rules_java_codestyle.html#unnecessaryboxing)

**Description:** PMD reports "Unnecessary explicit conversion from Integer to Long" violation, but removing the explicit conversion will break compilation.

**Code Sample demonstrating the issue:**

```java
@SuppressWarnings({"PMD.SystemPrintln", "PMD.NoPackage"})
public final class UnnecessaryBoxing {

  private UnnecessaryBoxing() {
  }

  public static void addLong(Long parameter) {
    System.out.println("parameter = " + parameter);
  }

  public static Integer getValue() {
    return 42;
  }

  public static void main(String[] args) {
    addLong(Long.valueOf(getValue())); // PMD complains: "Unnecessary explicit conversion from Integer to Long"
    //addLong(getValue()); // Does not compile: "incompatible types: Integer cannot be converted to Long"
  }
}
```

**Expected outcome:**

PMD reports a violation at line 16 (see comment), but that's wrong. That's a false positive.

**Running PMD through:** *Gradle*

**Comment by oowekyala** [comment]
That's because your code is doing the conversion chain `Integer ~> int ~> long -> Long`, where the first two are implicit. You can "simplify" that by writing `getValue().longValue()` and not call `Long.valueOf`. That way you do `Integer -> long ~> Long`. I haven't tested that this satisfies the rule but I believe it would.

**Comment by andygoossens** [comment]
I see. Using `addLong(getValue().longValue());` in my example would indeed have satisfied the rule.

**Comment by julius-d** [comment]
Hi!
I am currently trying to update to PMD 7.
And I am also having a hard time finding out how to make the code pass the `UnnecessaryBoxing` check.
Currently I am believing that PMD 7 reports a false positive to me:

```java
import java.util.Optional;

public class Example {

  record Item(int cents) {}

  Long map(Item item) {
    return Optional.ofNullable(item).map(it -> Long.valueOf(it.cents())).orElse(null);
  }
}
```
PMD output
```
PMD Failure: Example:10 Rule:UnnecessaryBoxing Priority:3 
Unnecessary explicit conversion from int to Function<? super Item, ? extends Long>
```

Is it a false positive?

**Comment by oowekyala** [comment]
Yes this is a false positive, can you report a new issue? FWIW this can be fixed by writing `map(it -> (long) it.cents())`

**Comment by julius-d** [comment]
I created #4924 for it.
